zitadel/v2/sms_provider_twilio/funcs.go

```diff
@@ -0,0 +1,127 @@
+package sms_provider_twilio
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/zitadel/zitadel-go/v2/pkg/client/zitadel/admin"
+
+	"github.com/zitadel/terraform-provider-zitadel/zitadel/v2/helper"
+)
+
+func delete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	tflog.Info(ctx, "started delete")
+
+	clientinfo, ok := m.(*helper.ClientInfo)
+	if !ok {
+		return diag.Errorf("failed to get client")
+	}
+
+	client, err := helper.GetAdminClient(clientinfo)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	_, err = client.RemoveSMSProvider(ctx, &admin.RemoveSMSProviderRequest{Id: d.Id()})
+	if err != nil {
+		return diag.Errorf("failed to delete sms provider twilio: %v", err)
+	}
+	return nil
+}
+
+func create(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	tflog.Info(ctx, "started create")
+
+	clientinfo, ok := m.(*helper.ClientInfo)
+	if !ok {
+		return diag.Errorf("failed to get client")
+	}
+
+	client, err := helper.GetAdminClient(clientinfo)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	resp, err := client.AddSMSProviderTwilio(ctx, &admin.AddSMSProviderTwilioRequest{
+		Sid:          d.Get(sidVar).(string),
+		Token:        d.Get(tokenVar).(string),
+		SenderNumber: d.Get(senderNumberVar).(string),
+	})
+	if err != nil {
+		return diag.Errorf("failed to create sms provider twilio: %v", err)
+	}
+	d.SetId(resp.Id)
+
+	return nil
+}
+
+func update(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	tflog.Info(ctx, "started update")
+
+	clientinfo, ok := m.(*helper.ClientInfo)
+	if !ok {
+		return diag.Errorf("failed to get client")
+	}
+
+	client, err := helper.GetAdminClient(clientinfo)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	sms, err := client.GetSMSProvider(ctx, &admin.GetSMSProviderRequest{})
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	sid := d.Get(sidVar).(string)
+	senderNumber := d.Get(senderNumberVar).(string)
+	twilio := sms.Config.GetTwilio()
+	if twilio.Sid != sid ||
+		twilio.SenderNumber != senderNumber {
+		_, err = client.UpdateSMSProviderTwilio(ctx, &admin.UpdateSMSProviderTwilioRequest{
+			Id:           d.Id(),
+			Sid:          sid,
+			SenderNumber: senderNumber,
+		})
+		if err != nil {
+			return diag.Errorf("failed to update sms provider twilio: %v", err)
+		}
+	} else {
+		_, err = client.UpdateSMSProviderTwilioToken(ctx, &admin.UpdateSMSProviderTwilioTokenRequest{
+			Id:    d.Id(),
+			Token: d.Get(tokenVar).(string),
+		})
+		if err != nil {
+			return diag.Errorf("failed to update sms provider twilio: %v", err)
+		}
+	}
```

Comment:
token update and sid update should be doable in a single run, imo

Reply:
problem is the nothing changed errors, which make the implementation not better in any way

Reply:
ok, I understand the problem now.

Did you consider using a separate "update_password" boolean property to make it explicit?

Reply:
Then the bool would be in the terraform status, the current problem is that I have no way of getting the information of the password to compare it to the resource attribute, I still have to set the password otherwise the functionality is not complete.
IMO, as long as the functionality is available, even if you have to apply it twice, it shouldn't be a problem. Besides with the boolean you have the same "nothing changed" error if you didn't change the password, but you have to terraform apply to change the boolean in the status to "false" if you just want to change the SID.